daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1446,37 +1492,41 @@ private[lf] object SBuiltin {
     val templateId: TypeConName
 
     // Callback from the engine returned NotFound
-    def handleKeyFound(machine: Machine, cid: V.ContractId): Unit
+    def handleKeyFound(machine: Machine, cid: V.ContractId): Control
     // We already saw this key, but it was undefined or was archived
-    def handleKeyNotFound(machine: Machine, gkey: GlobalKey): Boolean
+    def handleKeyNotFound(machine: Machine, gkey: GlobalKey): (Control, Boolean)
```

Comment:
it's orthogonal to this PR but trying to understand the `Boolean` here made my head hurt, and I'm not sure it's even used anywhere other than [here](https://github.com/digital-asset/daml/blob/main/daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala#L261-L266), and [the code below](https://github.com/digital-asset/daml/blob/main/daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala#L283-L288) explicitly discards it. Just something we could simplify later.

Reply:
perhaps returning an `Option[Control]` would make more sense

Reply:
I don't have a good knowledge of the details here. (didn't need them for the refactoring)
But I also found this `Bool` rather odd.
Thanks for digging in... and seeing it may be redundant.
If so , I totally agree we should kill it with fire in a subsequent clean up.

Reply:
Moises and I dug deeper into the `Bool` returned by `KeyOperation.handleKeyNotFound`
Recording out findings here...

It winds up as the return of the `callback` of `SResultNeedKey` in `SResult.scala`

The only place which checks this bool is the function `missingWith` in `ScenarioRunner`
This calls the `callback` with `None`, and will error if the returned boolean is `false`.
(When it calls the `callback` with `Some`, the bool is discarded
Also, the caller in `Engine.scala` discards the bool completely.)

The only place which returns `false` is `KeyOperation.Fetch.handleKeyNotFound`

If the check in `missingWith` is disabled, then following test fails:
```
$ bazel run //compiler/damlc/tests:integration-v1dev -- -p ContractKeyNotEffective
```

Daml code In:
```
daml/compiler/damlc/tests/daml-test-files/ContractKeyNotEffective.daml
```

Which seems to be checking the ledger behaviour when we go back in time...
```
  pass (- seconds 1)
```

What is this about?

Reply:
Adding to the information above:
- `KeyOperation.handleKeyNotFound` is used by `SBUFetchKey` and `SBULookupKey` in their respective `execute` methods
- it accounts for some of their subtle behavioural differences.

Reply:
Ah, the test failure for `bazel run //compiler/damlc/tests:integration-v1dev -- -p ContractKeyNotEffective` (aka `compiler/damlc/tests/daml-test-files/ContractKeyNotEffective.daml`) turned out to be a mismatched diagnostic. The test expects an error message

```
-- expected
contract <cid> not effective, but we found its key
```

but with this change we throw 

```
-- actual
Attempt to fetch or exercise by key but no contract with that key was found.
```

so really the boolean seems to be toggling whether execution after `missingWith`
* (true) continues normally (e.g. `lookupByKey` returns a `None`)
* (false) fails with one of the exceptions given to `missingWith` 

Perhaps this code would be easier to follow if `handleKeyNotFound` also took an exception as an argument, throwing it (I mean, purely returning a `Control.Expression(e)` where `e` is the exception) in the case of `KeyOperation.Fetch` and discarding it in `KeyOperation.Lookup`

Reply:
Just a further thought...
there is someting very odd about having a `Bool` return from the `SResultNeedKey` `callback`.

Every other `SResult*` which represents a question from the speedy-engine to the encompasing ledger has the form:
```
  final case class SResultSomeQuestion(
      questionParameters : QuestionParams,
      callback: Answer => Unit,
  ) extends SResult
```

But for `SResultNeedKey`, we have this extra `Bool`...
```
  final case class SResultNeedKey(
      key: GlobalKeyWithMaintainers,
      committers: Set[Party],
      callback: Option[ContractId] => Boolean,
  ) extends SResult
```
So, the engine asks the ledger for some information...
the ledger replies to the engine with an answer...
and then the engine tells the ledger what it thinks of the answer !

Reply:
And then based on what the engine thought of the answer, the ledger can follow a different path.

Reply:
exactly! not very continuation-y if you ask me :) I think this matches my intuition that the exception should be an argument to the callback, then the engine can decide if it throws that, something else, or continues execution

Reply:
yes, that would be much better

Reply:
And really it is just passing the information that the info is wanted by a Fetch or Lookup.

Reply:
yup! I can start working on a PR for this if that's fine by you :)

Reply:
:+